lib/mongoid/extensions/big_decimal.rb

```diff
@@ -46,7 +50,13 @@ module ClassMethods
         #
         # @return [ BigDecimal, nil ] A BigDecimal derived from the object or nil.
         def demongoize(object)
-          object && object.numeric? ? BigDecimal(object.to_s) : nil
+          if object
+            if object.is_a?(BSON::Decimal128)
+              object.to_big_decimal
+            elsif object.numeric?
```

Comment:
I would like the `numeric?` condition to be dropped and this path be taken for non-numeric input also. I see that non-numeric input was already trashed previously but I think this is not good behavior.

Test with a time field:

```
  class Foo
    include Mongoid::Document

    field :a
  end

  Foo.delete_all

  Foo.create!(a: 'zz')

  class Foo
    field :a, type: BigDecimal
  end

  p Foo.first, Foo.first.a
```

produces:

```
NoMethodError: undefined method `getlocal' for "zz":String
```

Test with the code as proposed with BigDecimal field:

```
  class Foo
    include Mongoid::Document

    field :a
  end

  Foo.delete_all

  Foo.create!(a: 'zz')

  class Foo
    field :a, type: Time
  end

  p Foo.first, Foo.first.a
```

produces:

```
#<Foo _id: 61d76a95a15d5d4e9c038186, a: "zz">
nil
```

I think it would be better to always attempt to construct a BigDecimal, allowing *that* call to raise whatever exceptions on problems.

@comandeo  thoughts?

Reply:
I disagree. I think demongoize should return nil unless the object can be represented as a BigDecimal. Should check the behavior of demongoize on other classes.

Reply:
@p-mongo I kind of agree with @johnnyshields here... Here are some examples of invalid values to demongoize:
```
irb(main):004:0> Float.demongoize("1.2asda")
=> 0.0
irb(main):005:0> Integer.demongoize("1234rqwdd")
=> 0
```
Now, I'm not sure why `0` is the right return value here but the important thing is that it doesn't raise an error. Let me know what you think...

Reply:
@Neilshweky is correct should make this return `0`. Important thing is that it doesn't raise error.

Reply:
Actually after rereading your comment, I agree @p-mongo. Let's let BigDecimal throw an error

Reply:
Wait--`demongoize` should NOT throw an error. Can you find examples of other classes where demongoize throws an error?

Imagine the user experience for this this... DB object can't be read... nightmare.

Reply:
@p-mongo @comandeo what do you think?

Reply:
I am inclined to agree with @johnnyshields here, we should not fail if we cannot demongoize value.

Imagine we have the following document:
```ruby
class Item
  include Mongoid::Document
  
  field :name, type: String
  field :kind, type: String
  field :amount, type: BigDecimal
end
```

Let's say our database contains following records:
```
[
  {
    _id: ObjectId("61d8889af7b3f3f5d961f974"),
    name: 'name1',
    kind: 'x',
    amount: Decimal128("1")
  },
  {
    _id: ObjectId("61d8889af7b3f3f5d961f975"),
    name: 'name2',
    kind: 'x',
    amount: 'error: bad input, cannot parse provided data'
  },
  {
    _id: ObjectId("61d8889af7b3f3f5d961f976"),
    name: 'name3',
    kind: 'y',
    amount: Decimal128("2")
  }
]
```

This is a legitimate situation. It can happen, for example, if there is another application that writes to the same collection, and something went wrong. And even if something went wrong, we still may want to work with the data.

Now we try to query the data:
```ruby
# A dumb example to illustrate the problem.
Item.where(kind: 'x').map do |i|
  if i.amount.nil?
    "<p>BAD RECORD</p>" # Or whatever I want to to do handle the problem.
  else
    css_class = if i.amount > 0
                  'green'
                else
                  'red'
                end
    "<p class='#{css_class}'>#{i.name}: #{i.amount}</p>"
  end  
end
```
Proposed solution will throw an exception in the middle of execution. Returning `nil` will give a user a chance to handle broken data.

As a user, I would probably expect that the query returns me requested data even if some fields cannot be deserialized. Setting the problematic field to nil seems to be a good indication of a problem to me, and the problem can be handled on application layer.

I do not think that we should return 0 here, though. Zero is a valid value.

Reply:
okay, I think that's a fair solution.

Reply:
I'm fine with either `nil` or `0`, but again would urge to look at what other classes do an and **keep it consistent**. (If there needs to be a follow-up PR to change BigDecimal + other classes uniformly, that's also fine...)

Reply:
I am afraid we are not consistent already:
```
irb(main):015:0> Date.demongoize('i am so bad')
date.rb:48:in `demongoize': undefined method `year' for "i am so bad":String (NoMethodError)

irb(main):018:0> Time.demongoize('so do i')
time.rb:57:in `demongoize': undefined method `getlocal' for "so do i":String (NoMethodError)

irb(main):020:0> DateTime.demongoize('i am also bad')
time.rb:57:in `demongoize': undefined method `getlocal' for "i am also bad":String (NoMethodError)

irb(main):021:0> Float.demongoize('i am sooooooo bad')
=> 0.0

irb(main):022:0> Integer.demongoize('could not be worse')
=> 0

irb(main):023:0> Range.demongoize('and yet it can')
=> nil..

irb(main):024:0> Set.demongoize('i have no idea what i am doing')
ruby/3.1.0/set.rb:280:in `do_with_enum': value must be enumerable (ArgumentError)

irb(main):026:0> Regexp.demongoize(42).class.name
=> "Integer"

irb(main):028:0> Hash.demongoize('i give up')
=> "i give up"
```
We do have to make all these consistent later, that is true.

Reply:
Since the two existing numeric types (`Float` and `Integer`) currently return `0`, I would be inclined to say `BigDecimal` should return `0`. Fine to raise a follow-up ticket/PR if we wish to change it later.

Reply:
I think for now it's better practice to return nil, and we should file a ticket to follow upon Float and Integer.

Reply:
I created https://jira.mongodb.org/browse/MONGOID-5221 to address the demongoization issue across all types.

Reply:
I am unresolving this discussion thread because it's referenced from https://jira.mongodb.org/browse/MONGOID-5221.

Reply:
I've crossed out my earlier comment here--BigDecimal correctly returns `nil` today for non-numeric string, we should do the same for Integer/Float. It's ready to merge in PR: https://github.com/mongodb/mongoid/pull/5220